Storage/StorageManager.php

```diff
@@ -26,58 +25,91 @@ class StorageManager
     public function __construct(FileLocatorInterface $fileLocator, iterable $adapters, iterable $cacheAdapters)
     {
         $this->fileLocator = $fileLocator;
-        $this->adapters = $adapters;
+
+        foreach ($adapters as $adapters) {
+            $this->adapters[] = $adapters;
+        }
 
         foreach ($cacheAdapters as $cacheAdapter) {
-            if (!$cacheAdapter instanceof CacheAdapterInterface) {
-                throw new \InvalidArgumentException(sprintf('Adapter %s can not be used for caching', get_class($cacheAdapter)));
-            }
+            $this->cacheAdapters[] = $cacheAdapter;
         }
 
-        $this->cacheAdapters = $cacheAdapters;
     }
 
     /**
-     * @return AdapterInterface[]|iterable
+     * @return StorageInterface[]|iterable
      */
     public function getAdapters()
     {
         return $this->adapters;
     }
 
+
+    /**
+     * @param StorageInterface $storageAdapter
+     * @return StorageManager
+     */
+    public function addAdapter(StorageInterface $storageAdapter) {
+
+        $this->adapters[] = $storageAdapter;
+        if($storageAdapter->supportCacheStore())
+        {
+            $this->cacheAdapters[] = $storageAdapter;
+        }
+        return $this;
+    }
+
     /**
      * @param string      $hash
      * @param string|null $context
      *
-     * @return string
+     * @return resource
      */
-    public function getFile(string $hash, ?string $context = null): string
+    public function getResource(string $hash, ?string $context = null)
     {
         return $this->read($this->adapters, $hash, $context);
     }
 
     /**
+     * @deprecated
      * @param string      $hash
      * @param string|null $context
      *
      * @return string
      */
-    public function getCacheFile(string $hash, ?string $context = null): string
+    public function getFile(string $hash, ?string $context = null): string
     {
-        return $this->read($this->cacheAdapters, $hash, $context);
+        $resource = $this->read($this->adapters, $hash, $context);
+        $filename = tempnam(sys_get_temp_dir(), 'EMS');
+        file_put_contents($filename, $resource);
+        return $filename;
```

Comment:
How temporary is your file ?

Consider http://php.net/manual/en/class.spltempfileobject.php